File: tutorial/T05-Param.cpp
/***************************
@Author: Chunel
@Contact: chunel@foxmail.com
@File: T05-Param.cpp
@Time: 2021/6/13 3:31 下午
@Desc: 本例主要演示，参数传递功能
***************************/

#include "MyGNode/MyWriteParamNode.h"
#include "MyGNode/MyReadParamNode.h"

using namespace CGraph;

void tutorial_param() {
    GPipelinePtr pipeline = GPipelineFactory::create();
    CStatus status;
    GElementPtr a, b, c, d, e, f = nullptr;

    status += pipeline->registerGElement<MyReadParamNode>(&a, {}, "readNodeA");     // 读取param中的信息，不做修改
    status += pipeline->registerGElement<MyReadParamNode>(&b, {a}, "readNodeB");
    status += pipeline->registerGElement<MyWriteParamNode>(&c, {a}, "writeNodeC");
    status += pipeline->registerGElement<MyWriteParamNode>(&d, {a}, "writeNodeD", 2);    // 对param中的iValue值+1，循环执行2次
    status += pipeline->registerGElement<MyReadParamNode>(&e, {a}, "readNodeE");
    status += pipeline->registerGElement<MyWriteParamNode>(&f, {b, c, d, e}, "writeNodeF");
    if (!status.isOK()) {
        return;    // 使用时，请对所有CGraph接口的返回值做判定。本例子中省略
    }

    status = pipeline->init();

    for (int i = 0; i < 3; i++) {
        status = pipeline->run();
        std::cout << "[CGraph] tutorial_param, loop : " << i + 1 << ", and run status = " << status.getCode() << std::endl;
    }

    status = pipeline->destroy();
    GPipelineFactory::remove(pipeline);
}


int main() {
    tutorial_param();
    return 0;
}

File: python/tutorial/T05-Param.py
"""
@Author: Chunel
@Contact: chunel@foxmail.com
@File: T05-Param
@Time: 2025/2/27 23:41
@Desc: 
"""

from pycgraph import GNode, GPipeline

from MyGNode.MyWriteParamNode import MyWriteParamNode
from MyGNode.MyReadParamNode import MyReadParamNode


def tutorial_param():
    pipeline = GPipeline()
    a, b, e = MyReadParamNode(), MyReadParamNode(), MyReadParamNode()
    c, d, f = MyWriteParamNode(), MyWriteParamNode(), MyWriteParamNode()

    pipeline.registerGElement(a, set(), "readNodeA")
    pipeline.registerGElement(b, {a}, "readNodeB")
    pipeline.registerGElement(c, {a}, "writeNodeC")
    pipeline.registerGElement(d, {a}, "writeNodeD", 2)
    pipeline.registerGElement(e, {a}, "readNodeE")
    pipeline.registerGElement(f, {b, c, d, e}, "writeNodeF")

    pipeline.init()

    for i in range(0, 3):
        status = pipeline.run()
        print('---- tutorial_param, loop : {0}, and run status = {1}'.format(i + 1, status.getCode()))

    pipeline.destroy()


if __name__ == '__main__':
    tutorial_param()
